File: commonrv/src/main/java/com/even/commonrv/decoration/StickyItemDecoration.java
package com.even.commonrv.decoration;

import android.graphics.Canvas;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.view.ViewGroup;

/**
 * @author Created by Even on 2019/4/22
 * Email: emailtopan@163.com
 */
public class StickyItemDecoration extends RecyclerView.ItemDecoration {
    /**
     * RecyclerView的adapter
     */
    private RecyclerView.Adapter<RecyclerView.ViewHolder> mAdapter;
    /**
     * RecyclerView的ViewHolder
     */
    private RecyclerView.ViewHolder mViewHolder;
    /**
     * 标题视图
     */
    private View mStickyItemView;
    /**
     * 标题距离顶部的距离
     */
    private int mStickyItemViewMarginTop;
    /**
     * 标题布局高度
     */
    private int mItemViewHeight;
    /**
     * 绑定数据的position
     */
    private int mBindDataPosition = -1;
    /**
     * 布局管理器
     */
    private LinearLayoutManager mLayoutManager;
    /**
     * 这里的ViewType必须和Adapter中的相同，默认为0
     */
    private int mViewType = 0;

    public StickyItemDecoration() {
    }

    public StickyItemDecoration(int mViewType) {
        this.mViewType = mViewType;
    }

    @Override
    public void onDrawOver(@NonNull Canvas c, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        super.onDrawOver(c, parent, state);
        //非空判断
        if (parent == null || parent.getAdapter() == null || parent.getAdapter().getItemCount() <= 0) {
            return;
        }
        if (mLayoutManager == null) {
            mLayoutManager = (LinearLayoutManager) parent.getLayoutManager();
        }
        getStickyViewHolder(parent);

        int itemCount = parent.getChildCount();
        for (int i = 0; i < itemCount; i++) {
            View itemView = parent.getChildAt(i);
            //如果是标题
            if ((boolean) itemView.getTag()) {
                if (itemView.getTop() <= 0) {
                    //标题布局全部显示效果
                    bindDataForStickyView(mLayoutManager.findFirstVisibleItemPosition(), parent.getMeasuredWidth());
                } else {
                    bindDataForStickyView(mLayoutManager.findFirstVisibleItemPosition() - 1, parent.getMeasuredWidth());
                }
                if (itemView.getTop() > 0 && itemView.getTop() <= mItemViewHeight) {
                    //处理两个标题重叠绘制效果
                    mStickyItemViewMarginTop = mItemViewHeight - itemView.getTop();
                } else {
                    mStickyItemViewMarginTop = 0;
                    //获取下一个标题
                    View nextStickyView = getNextStickyView(parent);
                    if (nextStickyView != null && nextStickyView.getTop() <= mItemViewHeight) {
                        //若两标题叠在一起了
                        //第二个标题盖住第一个标题多少了
                        mStickyItemViewMarginTop = mItemViewHeight - nextStickyView.getTop();
                    }
                }
                // 准备工作已就绪，开始画出吸附的标题
                drawStickyItemView(c);
                break;
            }
        }

    }

    /**
     * 给StickyView绑定数据
     *
     * @param position
     */
    private void bindDataForStickyView(int position, int width) {
        if (mViewHolder == null) {
            return;
        }
        //已经是吸附位置了 或 视图不存在
        mBindDataPosition = position;
//        //改变标题的展示效果，该方法在Adapter中
        mAdapter.onBindViewHolder(mViewHolder, mBindDataPosition);
        //设置布局位置及大小
        measureLayoutStickyItemView(width);
        //计算标题布局高度
        mItemViewHeight = mViewHolder.itemView.getBottom() - mViewHolder.itemView.getTop();
    }

    /**
     * 设置布局位置及大小
     *
     * @param parentWidth 父布局宽度
     */
    private void measureLayoutStickyItemView(int parentWidth) {
        if (mStickyItemView == null || !mStickyItemView.isLayoutRequested()) {
            return;
        }
        int widthSpec = View.MeasureSpec.makeMeasureSpec(parentWidth, View.MeasureSpec.EXACTLY);
        int heightSpec;

        ViewGroup.LayoutParams layoutParams = mStickyItemView.getLayoutParams();
        if (layoutParams != null && layoutParams.height > 0) {
            heightSpec = View.MeasureSpec.makeMeasureSpec(layoutParams.height, View.MeasureSpec.EXACTLY);
        } else {
            heightSpec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED);
        }

        mStickyItemView.measure(widthSpec, heightSpec);
        /**
         * mContentView.layout(l,t,r,b) ; 子布局相对于父布局的绘制的位置及大小。
         * l 和 t 是控件左边缘和上边缘相对于父类控件左边缘和上边缘的距离。r 和 b是控件右边缘和下边缘相对于父类控件左边缘和上边缘的距离。
         */
        mStickyItemView.layout(0, 0, mStickyItemView.getMeasuredWidth(), mStickyItemView.getMeasuredHeight());
    }

    /**
     * 得到下一个标题
     *
     * @param parent
     * @return
     */
    private View getNextStickyView(RecyclerView parent) {
        int num = 0;
        View nextStickyView = null;
        for (int m = 0, size = parent.getChildCount(); m < size; m++) {
            //循环获取每个子项
            View view = parent.getChildAt(m);
            if ((boolean) view.getTag() == true) {
                //拿到标题
                nextStickyView = view;
                num++;
            }
            //拿到第二个标题 ，就结束循环。
            if (num == 2) {
                break;
            }
        }
        return nextStickyView;
    }

    /**
     * 绘制标题
     *
     * @param canvas
     */
    private void drawStickyItemView(Canvas canvas) {
        if (mStickyItemView == null) {
            return;
        }
        //保存当前图层
        int saveCount = canvas.save();
        //图层转换位移
        canvas.translate(0, -mStickyItemViewMarginTop);
        mStickyItemView.draw(canvas);
        //恢复指定层的图层
        canvas.restoreToCount(saveCount);
    }

    /**
     * 得到标题的 viewHolder
     *
     * @param recyclerView
     */
    private void getStickyViewHolder(RecyclerView recyclerView) {
        //判断是否已创建
        if (mAdapter != null) {
            return;
        }
        mAdapter = recyclerView.getAdapter();
        //该方法属于Adapter中的重写Override
        mViewHolder = mAdapter.onCreateViewHolder(recyclerView, mViewType);
        //得到布局
        mStickyItemView = mViewHolder.itemView;
    }
}

File: commonrv/src/main/java/com/even/commonrv/decoration/StickyItemDecoration.kt
package com.even.commonrv.decoration

import android.graphics.Canvas
import android.view.View
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ItemDecoration

/**
 * @author Created by Even on 2019/4/22
 * Email: emailtopan@163.com
 */
class StickyItemDecoration : ItemDecoration {
    /**
     * 标题视图
     */
    private var mStickyItemView: View? = null

    private var mViewHolder: RecyclerView.ViewHolder? = null

    /**
     * 标题距离顶部的距离
     */
    private var mStickyItemViewMarginTop = 0

    /**
     * 标题布局高度
     */
    private var mItemViewHeight = 0

    /**
     * 绑定数据的position
     */
    private var mBindDataPosition = -1


    /**
     * 悬浮标题ViewType
     */
    private var mViewType = 0

    constructor(mViewType: Int) {
        this.mViewType = mViewType
    }

    constructor() {}

    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDrawOver(c, parent, state)
        parent.adapter?.let {
            val layoutManager = parent.layoutManager as LinearLayoutManager
            if (it.itemCount > 0) {
                if (mStickyItemView == null) {
                    mViewHolder = it.createViewHolder(parent, mViewType)
                    mStickyItemView = mViewHolder?.itemView
                }
                for (i in 0 until parent.childCount) {
                    val view = parent.getChildAt(i)
                    if (view.tag as Boolean) {
                        //标题
                        if (view.top <= 0) {
                            //标题布局全部显示效果
                            bindDataForStickyView(
                                it,
                                layoutManager.findFirstVisibleItemPosition(),
                                parent.measuredWidth
                            )
                        } else {
                            bindDataForStickyView(
                                it,
                                layoutManager.findFirstVisibleItemPosition() - 1,
                                parent.measuredWidth
                            )
                        }

                        if (view.top in 1..mItemViewHeight) {
                            //两个标题重叠
                            mStickyItemViewMarginTop = mItemViewHeight - view.top
                        } else {
                            mStickyItemViewMarginTop = 0
                            //获取下一个标题
                            val nextStickyView = getNextStickyView(parent)
                            if (nextStickyView != null && nextStickyView.top <= mItemViewHeight) {
                                //若两标题叠在一起了
                                //第二个标题盖住第一个标题多少了
                                mStickyItemViewMarginTop = mItemViewHeight - nextStickyView.top
                            }
                        }
                        drawStickyItemView(c)
                        break
                    }
                }

            }
        }
    }

    /**
     * 给StickyView绑定数据
     */
    private fun bindDataForStickyView(
        adapter: RecyclerView.Adapter<RecyclerView.ViewHolder>,
        position: Int,
        width: Int
    ) {
        //已经是吸附位置了 或 视图不存在
        mBindDataPosition = position
        adapter.onBindViewHolder(mViewHolder!!, mBindDataPosition)
        //设置布局位置及大小
        measureLayoutStickyItemView(width)
        //计算标题布局高度
        mItemViewHeight = mStickyItemView!!.bottom - mStickyItemView!!.top
    }

    /**
     * 设置布局位置及大小
     *
     * @param parentWidth 父布局宽度
     */
    private fun measureLayoutStickyItemView(parentWidth: Int) {
        if (mStickyItemView == null || !mStickyItemView!!.isLayoutRequested) {
            return
        }
        val widthSpec = View.MeasureSpec.makeMeasureSpec(
            parentWidth,
            View.MeasureSpec.EXACTLY
        )
        val heightSpec: Int
        val layoutParams = mStickyItemView!!.layoutParams
        heightSpec = if (layoutParams != null && layoutParams.height > 0) {
            View.MeasureSpec.makeMeasureSpec(
                layoutParams.height,
                View.MeasureSpec.EXACTLY
            )
        } else {
            View.MeasureSpec.makeMeasureSpec(
                0,
                View.MeasureSpec.UNSPECIFIED
            )
        }
        mStickyItemView!!.measure(widthSpec, heightSpec)
        mStickyItemView!!.layout(
            0,
            0,
            mStickyItemView!!.measuredWidth,
            mStickyItemView!!.measuredHeight
        )
    }


    /**
     * 得到下一个标题
     */
    private fun getNextStickyView(parent: RecyclerView): View? {
        var num = 0
        var nextStickyView: View? = null
        var m = 0
        while (m < parent.childCount) {
            //循环获取每个子项
            val view = parent.getChildAt(m)
            if (view.tag as Boolean) {
                //拿到标题
                nextStickyView = view
                num++
            }
            //拿到第二个标题 ，就结束循环。
            if (num == 2) {
                break
            }
            m++
        }
        return nextStickyView
    }

    /**
     * 绘制标题
     */
    private fun drawStickyItemView(canvas: Canvas) {
        //保存当前图层
        val saveCount = canvas.save()
        //图层转换位移
        canvas.translate(0f, -mStickyItemViewMarginTop.toFloat())
        mStickyItemView?.draw(canvas)
        //恢复指定层的图层
        canvas.restoreToCount(saveCount)
    }
}